### PR TITLE
feat(Tree): Adding exclusive flag

### PR DESCRIPTION
**Pull request by priyankar205**
Adding the exclusive flag. So that only one subtree can be open at a time

### CODE REVIEW

File: docs/src/examples/components/Tree/Types/TreeExclusiveExample.shorthand.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+const titleRenderer = (Component, { content, open, hasSubtree, ...restProps }) => (
+  <Component open={open} hasSubtree={hasSubtree} {...restProps}>
+    {hasSubtree && <Icon name={open ? 'arrow down' : 'arrow right'} />}
+    <span>{content}</span>
+  </Component>
+)
+
+const TreeExclusiveExample = () => (
+  <Tree items={items} renderItemTitle={titleRenderer} exclusive={true} />
```
Comment (mnajdova) [inline docs/src/examples/components/Tree/Types/TreeExclusiveExample.shorthand.tsx]: You don't need `exclusive={true}`, in the component, just exclusive is enough.
`<Tree items={items} renderItemTitle={titleRenderer} exclusive />`

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 2]
```diff
@@ -78,20 +90,24 @@ class TreeItem extends AutoControlledComponent<ReactProps<TreeItemProps>, TreeIt
...
+      this.setState({
         open: !this.state.open,
       })
+      _.invoke(this.props, 'onOpenChanged', e, this.props)
       _.invoke(predefinedProps, 'onClick', e, titleProps)
     },
   })
 
   renderContent() {
     const { items, title, renderItemTitle } = this.props
```
Comment (mnajdova) [inline packages/react/src/components/Tree/TreeItem.tsx]: `const { items, title, renderItemTitle, exclusive } = this.props`
and then use the `exclusive` it would be much more readable

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 3]
```diff
@@ -78,20 +90,24 @@ class TreeItem extends AutoControlledComponent<ReactProps<TreeItemProps>, TreeIt
...
-    const { open } = this.state
-
+    const open = this.props.exclusive ? this.props.open : this.state.open
```
Comment (mnajdova) [inline packages/react/src/components/Tree/TreeItem.tsx]: It looks like the open prop should be auto controlled prop, then you can make decision of when it should be changed. Even if it is auto controlled, you can control it from the parent component.

File: packages/react/src/components/Tree/TreeTitle.tsx [hunk 4]
```diff
@@ -62,10 +71,10 @@ class TreeTitle extends UIComponent<ReactProps<TreeTitleProps>> {
     return (
       <ElementType
         className={classes.root}
-        onClick={this.handleClick}
         {...accessibility.attributes.root}
         {...rtlTextContainer.getAttributes({ forElements: [children, content] })}
         {...unhandledProps}
+        onClick={this.handleClick}
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeTitle.tsx]:
```jsx
className={classes.root}
onClick={this.handleClick}
```

Please move it  and place it after the `className` prop

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 5]
```diff
@@ -52,21 +63,22 @@ export interface TreeItemState {
   open?: boolean
 }
 
-class TreeItem extends AutoControlledComponent<ReactProps<TreeItemProps>, TreeItemState> {
+class TreeItem extends UIComponent<ReactProps<TreeItemProps>, TreeItemState> {
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeItem.tsx]: Why this change is necessary? This means that controlled mode will not work **(!)**, see example:
https://codesandbox.io/s/845j354jx2?module=/example.js

Please revert it.

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 6]
```diff
@@ -52,21 +63,22 @@ export interface TreeItemState {
...
-
   static propTypes = {
     ...commonPropTypes.createCommon({
       content: false,
     }),
     defaultOpen: PropTypes.bool,
     items: customPropTypes.collectionShorthand,
+    index: PropTypes.number,
+    exclusive: PropTypes.bool,
+    onOpenChanged: PropTypes.func,
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeItem.tsx]:
```suggestion
    onOpenChange: PropTypes.func,
```

We should go with `onOpenChange` as it was defined in other components.

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 7]
```diff
@@ -78,20 +90,24 @@ class TreeItem extends AutoControlledComponent<ReactProps<TreeItemProps>, TreeIt
...
+  }
+
   handleTitleOverrides = predefinedProps => ({
     onClick: (e, titleProps) => {
       e.preventDefault()
-      this.trySetState({
+      this.setState({
         open: !this.state.open,
       })
+      _.invoke(this.props, 'onOpenChanged', e, this.props)
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeItem.tsx]: `onOpenChange` should be invoked only when it has sub items:

```jsx
if (hasSubtree) _.invoke(this.props, 'onOpenChanged', e, this.props)
```

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 8]
```diff
@@ -78,20 +90,24 @@ class TreeItem extends AutoControlledComponent<ReactProps<TreeItemProps>, TreeIt
     accessibility: defaultBehavior,
   }
 
+  public state = {
+    open: false,
+  }
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeItem.tsx]: As this component should be autocontrolled, please use `getInitialAutoControlledState()`

```jsx
  getInitialAutoControlledState({ multiple, search }: DropdownProps): DropdownState {
    return {
      open: false,
    }
  }
```

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 9]
```diff
@@ -52,21 +63,22 @@ export interface TreeItemState {
...
 
-  static autoControlledProps = ['open']
-
   static propTypes = {
     ...commonPropTypes.createCommon({
       content: false,
     }),
     defaultOpen: PropTypes.bool,
     items: customPropTypes.collectionShorthand,
+    index: PropTypes.number,
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeItem.tsx:80]: Please add `index` to `TreeItemProps`

File: packages/react/src/components/Tree/Tree.tsx [hunk 10]
```diff
@@ -50,24 +53,43 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
+  state = {
+    clickedIndex: -1,
+  }
+
   public static defaultProps = {
     as: 'ul',
     accessibility: defaultBehavior,
   }
 
+  handleTreeItemOverrides = predefinedProps => ({
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```suggestion
  handleTreeItemOverrides = (predefinedProps: TreeItemProps) => ({
```

File: packages/react/src/components/Tree/Tree.tsx [hunk 11]
```diff
@@ -50,24 +53,43 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
+    onOpenChanged: (e, treeItemProps) => {
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```suggestion
    onOpenChanged: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
```

File: packages/react/src/components/Tree/Tree.tsx [hunk 12]
```diff
@@ -50,24 +53,43 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
       content: false,
     }),
+    exclusive: PropTypes.bool,
     items: customPropTypes.collectionShorthand,
     renderItemTitle: PropTypes.func,
     rtlAttributes: PropTypes.func,
   }
 
+  state = {
+    clickedIndex: -1,
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]: Please use `activeIndex`

**Comment by layershifter** [comment]
This PR contains the chicken-egg problem as I mentioned originally in #479: we have control of `activeIndex`/`clickedIndex` in `Tree` and `open` in `TreeItem` 🤔 

I expect from API something like this:

```jsx
const items = [
  {
    key: '1',
    activeIndex: 0, // < ---
    title: 'one',
    items: [
      {
        key: '2',
        activeIndex: 0, // < ---
        title: 'one one',
        items: [
          {
            key: '3',
            title: 'one one one',
          },
        ],
      },
    ],
  },
  {
    key: '4',
    title: 'two',
    items: [
      {
        key: '5',
        title: 'two one',
      },
    ],
  },
]

<Tree items={items} activeIndex={1} />
```

I suggest to move all `activeIndex` functionality to `TreeItem` and use factory in `Tree`:

```jsx
render() {
  return TreeItem.create({ as, items })
}
```

@mnajdova what do you think?

File: packages/react/src/components/Tree/TreeItem.tsx (hunk 9 above)
Comment (priyankar205) [reply packages/react/src/components/Tree/TreeItem.tsx]: @layershifter What do you mean by this?

Comment (priyankar205) [reply packages/react/src/components/Tree/TreeItem.tsx:80]: Done

**Comment by mnajdova** [comment]
> This PR contains the chicken-egg problem as I mentioned originally in #479: we have control of activeIndex/clickedIndex in Tree and open in TreeItem 🤔

I agree with your point. I would restrain for suggesting what the changes should be (although I agree with you). The goal of it should be, it should be able to auto control it's state and allow the user to control it. With this changes I don't think we are following this pattern. @priyankar205 please consider the comments from @layershifter - let's make this component follow the stardust principles implemented in all other components.

**Comment by priyankar205** [comment]
@mnajdova @layershifter No other component supports nested subcomponents. I will try to make the suggested changes.

**Comment by priyankar205** [comment]
> This PR contains the chicken-egg problem as I mentioned originally in #479: we have control of `activeIndex`/`clickedIndex` in `Tree` and `open` in `TreeItem` 🤔
> 
> I expect from API something like this:
> 
> ```js-jsx
> const items = [
>   {
>     key: '1',
>     activeIndex: 0, // < ---
>     title: 'one',
>     items: [
>       {
>         key: '2',
>         activeIndex: 0, // < ---
>         title: 'one one',
>         items: [
>           {
>             key: '3',
>             title: 'one one one',
>           },
>         ],
>       },
>     ],
>   },
>   {
>     key: '4',
>     title: 'two',
>     items: [
>       {
>         key: '5',
>         title: 'two one',
>       },
>     ],
>   },
> ]
> 
> <Tree items={items} activeIndex={1} />
> ```
> 
> I suggest to move all `activeIndex` functionality to `TreeItem` and use factory in `Tree`:
> 
> ```js-jsx
> render() {
>   return TreeItem.create({ as, items })
> }
> ```
> 
> @mnajdova what do you think?

If I move activeIndex to the TreeItem then how do I control the sibling TreeItems. How do I close the rest when the exclusive flag is there

File: packages/react/src/components/Tree/Tree.tsx [hunk 13]
```diff
@@ -39,7 +48,7 @@ export interface TreeProps extends UIComponentProps, ChildrenComponentProps {
 /**
  * Allows users to display data organised in tree-hierarchy.
  */
-class Tree extends UIComponent<ReactProps<TreeProps>> {
+class Tree extends AutoControlledComponent<ReactProps<TreeProps>, any> {
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]: Let's create an interface for the state:

```tsx
export interface TreeState {
  activeIndex: number[] | number
}
```

```suggestion
class Tree extends AutoControlledComponent<ReactProps<TreeProps>, TreeState> {
```

File: packages/react/src/components/Tree/Tree.tsx [hunk 14]
```diff
@@ -60,14 +78,47 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
     accessibility: defaultBehavior,
   }
 
+  static autoControlledProps = ['activeIndex']
+
+  getInitialAutoControlledState({ defaultActiveIndex, exclusive }) {
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```suggestion
  getInitialAutoControlledState({ defaultActiveIndex, exclusive }): TreeState {
```

File: packages/react/src/components/Tree/Tree.tsx [hunk 15]
```diff
@@ -60,14 +78,47 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
+    }
+  }
+
+  computeNewIndex = index => {
+    const { activeIndex } = this.state
+    const { exclusive } = this.props
+
+    if (exclusive) return index === activeIndex ? -1 : index
+    // check to see if index is in array, and remove it, if not then add it
+    return _.includes(activeIndex, index) ? _.without(activeIndex, index) : [...activeIndex, index]
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```suggestion
    return _.includes(activeIndex as number[], index) ? _.without(activeIndex as number[], index) : [...activeIndex as number[], index]
```

As we apply checks on `exclusive`, TS doesn't know that `activeIndex` is `number`. @mnajdova do you have better ideas? 💭

File: packages/react/src/components/Tree/Tree.tsx [hunk 16]
```diff
@@ -60,14 +78,47 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
 
+  static autoControlledProps = ['activeIndex']
+
+  getInitialAutoControlledState({ defaultActiveIndex, exclusive }) {
+    return {
+      activeIndex: defaultActiveIndex === undefined ? (exclusive ? -1 : []) : defaultActiveIndex,
+    }
+  }
+
+  computeNewIndex = index => {
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```suggestion
  computeNewIndex = (index: number) => {
```

File: packages/react/src/components/Tree/Tree.tsx [hunk 17]
```diff
@@ -60,14 +78,47 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
+      _.invoke(predefinedProps, 'onOpenChange', e, treeItemProps)
+    },
+  })
+
   renderContent() {
-    const { items, renderItemTitle } = this.props
+    const { items, renderItemTitle, exclusive } = this.props
 
-    return _.map(items, item =>
+    return _.map(items, (item, index) =>
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```suggestion
    return _.map(items, (item: ShorthandValue, index: number) =>
```

To clarify types

File: packages/react/src/components/Tree/Tree.tsx [hunk 18]
```diff
@@ -60,14 +78,47 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
+      this.trySetState({
+        activeIndex: this.computeNewIndex(index),
+      })
+      _.invoke(predefinedProps, 'onOpenChange', e, treeItemProps)
+    },
+  })
+
   renderContent() {
-    const { items, renderItemTitle } = this.props
+    const { items, renderItemTitle, exclusive } = this.props
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx:118]:
```suggestion
    const { items, renderItemTitle, exclusive } = this.props
    const { activeIndex } = this.state
```

Let's desctruct `activeIndex `

File: packages/react/src/components/Tree/Tree.tsx [hunk 19]
```diff
@@ -60,14 +78,47 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
-    return _.map(items, item =>
+    return _.map(items, (item, index) =>
       TreeItem.create(item, {
         defaultProps: {
+          index,
+          exclusive,
           renderItemTitle,
+          open: exclusive
+            ? index === this.state.activeIndex
+            : _.includes(this.state.activeIndex, index),
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```tsx
open: exclusive ? index === activeIndex : _.includes(activeIndex as number[], index),
```

Now can be one-liner 👍

File: packages/react/src/components/Tree/Tree.tsx [hunk 20]
```diff
@@ -60,14 +78,47 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
         },
+        overrideProps: this.handleTreeItemOverrides,
       }),
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx:131]: Let's also add shorthand factory, we will use it in the `TreeItem`:

```diff
+Tree.create = createShorthandFactory({ Component: Tree, mappedArrayProp: 'items' })
+
export default Tree
```

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 21]
```diff
@@ -104,7 +113,13 @@ class TreeItem extends AutoControlledComponent<ReactProps<TreeItemProps>, TreeIt
...
           overrideProps: this.handleTitleOverrides,
         })}
-        {hasSubtree && open && <Tree items={items} renderItemTitle={renderItemTitle} />}
+        {hasSubtree && open && (
+          <Tree
+            items={items}
+            renderItemTitle={renderItemTitle}
+            exclusive={!!this.props.exclusive}
+          />
+        )}
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeItem.tsx]: Factory will handle falsy values for us, so we can just do this:
```tsx
{open && Tree.create(items, {
  defaultProps: {
    exclusive,
    renderItemTitle,
  },
})}
```

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 22]
```diff
@@ -78,20 +91,16 @@ class TreeItem extends AutoControlledComponent<ReactProps<TreeItemProps>, TreeIt
...
+      _.invoke(this.props, 'onOpenChange', e, this.props)
       _.invoke(predefinedProps, 'onClick', e, titleProps)
     },
   })
 
   renderContent() {
-    const { items, title, renderItemTitle } = this.props
-    const { open } = this.state
-
+    const { items, title, renderItemTitle, open } = this.props
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeItem.tsx]:
```suggestion
    const { exclusive, items, title, renderItemTitle, open } = this.props
```

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 23]
```diff
@@ -78,20 +91,16 @@ class TreeItem extends AutoControlledComponent<ReactProps<TreeItemProps>, TreeIt
...
   }
 
-  handleTitleOverrides = predefinedProps => ({
+  handleTitleOverrides = (predefinedProps: TreeItemProps) => ({
     onClick: (e, titleProps) => {
       e.preventDefault()
-      this.trySetState({
-        open: !this.state.open,
-      })
+      _.invoke(this.props, 'onOpenChange', e, this.props)
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeItem.tsx]: Actually, now this is `onTitleClick`

```suggestion
      _.invoke(this.props, 'onTitleClick', e, this.props)
```

File: packages/react/src/components/Tree/Tree.tsx [hunk 24]
```diff
@@ -60,14 +78,47 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
+    return _.includes(activeIndex, index) ? _.without(activeIndex, index) : [...activeIndex, index]
+  }
+
+  handleTreeItemOverrides = predefinedProps => ({
+    onOpenChange: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      const { index } = treeItemProps
+      this.trySetState({
+        activeIndex: this.computeNewIndex(index),
+      })
+      _.invoke(predefinedProps, 'onOpenChange', e, treeItemProps)
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```tsx
    onTitleClick: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
      this.trySetState({ activeIndex: this.computeNewIndex(treeItemProps.index) })
      _.invoke(predefinedProps, 'onTitleClick', e, treeItemProps)
    },
```

https://github.com/stardust-ui/react/pull/1018/files#r263292996

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 25]
```diff
@@ -67,6 +77,9 @@ class TreeItem extends AutoControlledComponent<ReactProps<TreeItemProps>, TreeIt
     }),
     defaultOpen: PropTypes.bool,
     items: customPropTypes.collectionShorthand,
+    index: PropTypes.number,
+    exclusive: PropTypes.bool,
+    onOpenChange: PropTypes.func,
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeItem.tsx]:
```suggestion
    onTitleClick: PropTypes.func,
```

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 26]
```diff
@@ -25,12 +30,21 @@ export interface TreeItemProps extends UIComponentProps, ChildrenComponentProps
...
   defaultOpen?: boolean
 
+  /** The index of the item among its sibbling */
+  index: number
+
   /** Array of props for sub tree. */
   items?: ShorthandValue[]
 
+  /** Callback for toggling the current tree item as open element in the menu. */
+  onOpenChange?: ComponentEventHandler<TreeItemProps>
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeItem.tsx]:
```suggestion
  onTitleClick?: ComponentEventHandler<TreeItemProps>
```

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 27]
```diff
@@ -25,12 +30,21 @@ export interface TreeItemProps extends UIComponentProps, ChildrenComponentProps
...
   /** Initial open value. */
   defaultOpen?: boolean
 
+  /** The index of the item among its sibbling */
+  index: number
+
   /** Array of props for sub tree. */
   items?: ShorthandValue[]
 
+  /** Callback for toggling the current tree item as open element in the menu. */
```
Comment (layershifter) [inline packages/react/src/components/Tree/TreeItem.tsx]:
```suggestion
  /** Called when a tree title is clicked. */
```

File: packages/react/src/components/Tree/Tree.tsx [hunk 28]
```diff
@@ -60,14 +78,47 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
+  computeNewIndex = index => {
+    const { activeIndex } = this.state
+    const { exclusive } = this.props
+
+    if (exclusive) return index === activeIndex ? -1 : index
+    // check to see if index is in array, and remove it, if not then add it
+    return _.includes(activeIndex, index) ? _.without(activeIndex, index) : [...activeIndex, index]
+  }
+
+  handleTreeItemOverrides = predefinedProps => ({
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```suggestion
  handleTreeItemOverrides = (predefinedProps: TreeItemProps) => ({
```

File: docs/src/examples/components/Tree/Types/index.tsx [hunk 29]
```diff
@@ -14,6 +14,11 @@ const Types = () => (
       description="A Tree with customized title rendering."
       examplePath="components/Tree/Types/TreeTitleCustomizationExample"
     />
+    <ComponentExample
+      title="Exclusive Flag"
```
Comment (layershifter) [inline docs/src/examples/components/Tree/Types/index.tsx]:
```suggestion
      title="Exclusive"
```

File: CHANGELOG.md [hunk 30]
```diff
@@ -20,6 +20,9 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 ### Features
 - Add `inline` prop in the `Popup` for rendering the content next to the trigger element @mnajdova ([#1017](https://github.com/stardust-ui/react/pull/1017))
 
+- Add `exclusive` prop in the `Tree` for expanding one tree item at a time
```
Comment (layershifter) [inline CHANGELOG.md:25]: Please remove empty line before, 22

File: packages/react/src/components/Tree/Tree.tsx [hunk 31]
```diff
@@ -60,14 +83,48 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
+      ? _.without(activeIndex as number[], index)
+      : [...(activeIndex as number[]), index]
+  }
+
+  handleTreeItemOverrides = (predefinedProps: TreeItemProps) => ({
+    onTitleClick: (e: React.SyntheticEvent, treeItemProps: TreeItemProps) => {
+      const { index } = treeItemProps
+      this.trySetState({
+        activeIndex: this.computeNewIndex(index),
+      })
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```diff
-      const { index } = treeItemProps
-      this.trySetState({
-        activeIndex: this.computeNewIndex(index),
-      })
 this.trySetState({ activeIndex: this.computeNewIndex(treeItemProps.index) })
```

nit: 4 lines vs 1

**Review by layershifter — APPROVED** [review]

File: packages/react/src/components/Tree/Tree.tsx [hunk 32]
```diff
@@ -60,14 +83,45 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
     accessibility: defaultBehavior,
   }
 
+  static autoControlledProps = ['activeIndex']
```
Comment (mnajdova) [inline packages/react/src/components/Tree/Tree.tsx:86]: 👍

File: packages/react/src/components/Tree/Tree.tsx [hunk 33]
```diff
@@ -60,14 +83,45 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
+    const { activeIndex } = this.state
 
-    return _.map(items, item =>
+    return _.map(items, (item: ShorthandValue, index: number) =>
       TreeItem.create(item, {
         defaultProps: {
+          index,
+          exclusive,
           renderItemTitle,
+          open: exclusive ? index === activeIndex : _.includes(activeIndex as number[], index),
```
Comment (mnajdova) [inline packages/react/src/components/Tree/Tree.tsx]: Normalize the `activeIndex` as an array before doing this checks. If I don't use the exclusive, but use just defaultActiveIndex={3}, this may break. Do something like:
```
const activeIndexAsArray = _.isArray(activeIndex) ? activeIndex : [activeIndex]
```

and then always use the `_.include`

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 34]
```diff
@@ -78,20 +91,16 @@ class TreeItem extends AutoControlledComponent<ReactProps<TreeItemProps>, TreeIt
     accessibility: defaultBehavior,
   }
 
-  handleTitleOverrides = predefinedProps => ({
+  handleTitleOverrides = (predefinedProps: TreeItemProps) => ({
```
Comment (mnajdova) [inline packages/react/src/components/Tree/TreeItem.tsx]: Isn't the `predefinedProps` of type `TreeTitleProps`? @layershifter

Comment (layershifter) [reply packages/react/src/components/Tree/TreeItem.tsx]: Yep, my typo, great catch 👍

File: packages/react/src/components/Tree/Tree.tsx (hunk 33 above)
Comment (priyankar205) [reply packages/react/src/components/Tree/Tree.tsx]: This is consistent with the accordion.
screen shot 2019-03-07 at 6 09 57 pm

Comment (mnajdova) [reply packages/react/src/components/Tree/Tree.tsx]: Still, the user may provide the example showed above. If this is how it is implemented the Accordion, it is another story and we may fix it there as well.

File: packages/react/src/components/Tree/Tree.tsx [hunk 35]
```diff
@@ -60,14 +83,53 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
 
+  static autoControlledProps = ['activeIndex']
+
+  getInitialAutoControlledState({ exclusive }): TreeState {
+    return {
+      activeIndex: exclusive ? -1 : [],
+    }
+  }
+
+  getActiveIndexAsArray() {
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```suggestion
  getActiveIndexAsArray(): number[] {
```

File: packages/react/src/components/Tree/Tree.tsx [hunk 36]
```diff
@@ -60,14 +83,53 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
+
+  computeNewIndex = (index: number) => {
+    const { exclusive } = this.props
+
+    if (exclusive) return index
+    const activeIndexAsArray = this.getActiveIndexAsArray()
+    // check to see if index is in array, and remove it, if not then add it
+    return _.includes(activeIndexAsArray as number[], index)
+      ? _.without(activeIndexAsArray as number[], index)
+      : [...(activeIndexAsArray as number[]), index]
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]: Now we don't need `as number`

File: packages/react/src/components/Tree/Tree.tsx [hunk 37]
```diff
@@ -60,14 +83,53 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
+      this.trySetState({ activeIndex: this.computeNewIndex(treeItemProps.index) })
+      _.invoke(predefinedProps, 'onTitleClick', e, treeItemProps)
+    },
+  })
+
   renderContent() {
-    const { items, renderItemTitle } = this.props
+    const { items, renderItemTitle, exclusive } = this.props
+    const { activeIndex } = this.state
+    const activeIndexAsArray = this.getActiveIndexAsArray()
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```diff
-const { activeIndex } = this.state
-const activeIndexAsArray = this.getActiveIndexAsArray()
+const activeIndexes = this.getActiveIndexAsArray()
```

File: packages/react/src/components/Tree/Tree.tsx [hunk 38]
```diff
@@ -60,14 +83,53 @@ class Tree extends UIComponent<ReactProps<TreeProps>> {
...
-    return _.map(items, item =>
+    return _.map(items, (item: ShorthandValue, index: number) =>
       TreeItem.create(item, {
         defaultProps: {
+          index,
+          exclusive,
           renderItemTitle,
+          open: exclusive
+            ? index === activeIndex
+            : _.includes(activeIndexAsArray as number[], index),
```
Comment (layershifter) [inline packages/react/src/components/Tree/Tree.tsx]:
```diff
-          open: exclusive
-            ? index === activeIndex
-            : _.includes(activeIndexAsArray as number[], index),
+         open  : _.includes(activeIndexes, index),

File: packages/react/src/components/Tree/Tree.tsx (hunk 35 above)
Comment (layershifter) [reply packages/react/src/components/Tree/Tree.tsx]: I also suggest to rename to `getActiveIndexes()`

File: packages/react/src/components/Tree/Tree.tsx (hunk 38 above)
Comment (priyankar205) [reply packages/react/src/components/Tree/Tree.tsx]: For exclusive activeIndex is not an array. So this condition needs to be there.

Comment (layershifter) [reply packages/react/src/components/Tree/Tree.tsx]: `this.getActiveIndexes()` will always return array, even with `exclusive=true`
